central/networkgraph/service/service_impl.go

```diff
@@ -309,9 +312,68 @@ func (s *serviceImpl) getNetworkGraph(ctx context.Context, request *v1.NetworkGr
 			node.QueryMatch = true
 		}
 	}
+
+	if request.IncludePolicies {
+		err := s.enhanceWithNetworkPolicyIsolationInfo(ctx, graph)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to enhance graph nodes with Network Policy information")
+		}
+	}
+
 	return graph, nil
 }
 
+func (s *serviceImpl) enhanceWithNetworkPolicyIsolationInfo(ctx context.Context, graph *v1.NetworkGraph) error {
+	var deploymentIds []string
+	for _, node := range graph.GetNodes() {
+		if node.GetEntity().GetType() == storage.NetworkEntityInfo_DEPLOYMENT {
+			deploymentIds = append(deploymentIds, node.GetEntity().GetId())
+		}
+	}
+
+	// TODO(ROX-16312): Change this to a custom query once Postgres ships
+	deploymentObjects, err := s.deployments.GetDeployments(ctx, deploymentIds)
+	if err != nil {
+		return errors.Wrapf(err, "fetching deployment")
+	}
+
+	clusterNamespaceContext := set.NewSet[deploymentMatcher.ClusterNamespace]()
+	for _, deployment := range deploymentObjects {
+		clusterNamespaceContext.Add(deploymentMatcher.ClusterNamespace{
+			Cluster:   deployment.GetClusterId(),
+			Namespace: deployment.GetNamespace(),
+		})
+	}
+
+	matcher, err := deploymentMatcher.BuildMatcher(ctx, s.networkPolicy, clusterNamespaceContext)
+	if err != nil {
+		return errors.Wrap(err, "building deployment matcher")
+	}
+
+	isolationDetails := make(map[string]deploymentMatcher.IsolationDetails, len(deploymentObjects))
+	for _, deployment := range deploymentObjects {
+		details := matcher.GetIsolationDetails(deployment)
+		isolationDetails[deployment.GetId()] = deploymentMatcher.IsolationDetails{
+			IngressIsolated: details.IngressIsolated,
+			EgressIsolated:  details.EgressIsolated,
+			PolicyIDs:       details.PolicyIDs,
+		}
+	}
```

Comment:
I think you can avoid this loop if in the previous loop you create a `map[deploymentID]*storage.Deployment` and then in the following loop access this map instead of using the `isolationDetails` map. I think the memory usage should be similar because you are interchanging the new map with `isolationDetails` but this way you avoid one loop.

Reply:
Done